jsonrpc/server.go

```diff
@@ -262,52 +277,65 @@ func (s *Server) handleRequest(req *request) (*response, error) {
 }
 
 func buildArguments(params, handler any, configuredParams []Parameter) ([]reflect.Value, error) {
-	args := []reflect.Value{}
-	handlerType := reflect.TypeOf(handler)
-	paramCount := handlerType.NumIn()
-	paramsKind := reflect.Invalid
-	if !isNil(params) {
-		paramsKind = reflect.TypeOf(params).Kind()
+	var args []reflect.Value
+	if isNil(params) {
+		return args, nil
 	}
 
-	for idx := 0; idx < paramCount; idx++ {
-		valueContainer := reflect.New(handlerType.In(idx))
-		var requestValue any
-		found := true
+	handlerType := reflect.TypeOf(handler)
 
-		switch paramsKind {
-		case reflect.Slice:
-			paramsList := params.([]any)
-			if len(paramsList) != paramCount {
-				return nil, errors.New("missing param in list")
-			}
+	handlerParamValue := func(param any, t reflect.Type) (reflect.Value, error) {
+		handlerParam := reflect.New(t)
+		valueMarshaled, err := json.Marshal(param) // we have to marshal the value into JSON again
+		if err != nil {
+			return reflect.ValueOf(nil), err
+		}
+		err = json.Unmarshal(valueMarshaled, handlerParam.Interface())
+		if err != nil {
+			return reflect.ValueOf(nil), err
+		}
 
-			requestValue = paramsList[idx]
-		case reflect.Map:
-			paramsMap := params.(map[string]any)
-			requestValue, found = paramsMap[configuredParams[idx].Name]
-		case reflect.Invalid:
-			found = false
-			// nil params
-		default:
-			return nil, errors.New("impossible param type: check request.isSane")
+		return handlerParam.Elem(), nil
+	}
+
+	switch reflect.TypeOf(params).Kind() {
+	case reflect.Slice:
+		paramsList := params.([]any)
+
+		if len(paramsList) != handlerType.NumIn() {
+			return nil, errors.New("missing param in list")
 		}
 
-		if found {
-			valueMarshaled, err := json.Marshal(requestValue) // we have to marshal the value into JSON again
-			if err != nil {
-				return nil, err
-			}
-			err = json.Unmarshal(valueMarshaled, valueContainer.Interface())
+		for i, param := range paramsList {
+			v, err := handlerParamValue(param, handlerType.In(i))
 			if err != nil {
 				return nil, err
 			}
-		} else if !configuredParams[idx].Optional {
-			return nil, errors.New("missing non-optional param")
+			args = append(args, v)
 		}
+		return args, nil
+	case reflect.Map:
+		paramsMap := params.(map[string]any)
+
+		for i, configuredParam := range configuredParams {
+			// optional parameter
+			v := reflect.New(handlerType.In(i)).Elem()
```

Comment:
when named parameters are used, we are creating 2 instances for each parameter.

Reply:
I don't understand how. This line just creates a reflect value of the handler param type, how is it worse than `var v reflect.Value`?

Reply:
`reflect.Value` is a container that can encapsulate a value of any type.

`reflect.New` initializes a variable of the type that you pass to it, with that type's zero value. Then wraps it with a `reflect.Value`.   `var v reflect.Value` initializes the container only, not the value inside it. Container (`reflect.Value`) is a lightweight object, essentially a pointer with some metadata.

reflect.New might allocate on heap (twice, since we call it twice), `var v reflect.Value` allocates strictly on stack and since we eliminate a `reflect.New` call, we eliminate a heap allocation as well.

see `reflect.Value` definition; https://cs.opensource.google/go/go/+/refs/tags/go1.20.1:src/reflect/value.go;l=39

Reply:
Thanks for the explanation. 

>reflect.New allocates on heap (twice, since we call it twice), var v reflect.Value allocates strictly on stack and since we eliminate a reflect.New call, we eliminate a heap allocation as well.

Allocation depends on escape analysis, not on keywords. Since the second reference (in handlerParamValue) does not leave the scope of this function it will not be allocated on the heap.

Reply:
D had a great set of compile time features, we actually had a really similar construct that generated REST servers like we do with JSONRPC here. All allocations happened on stack with the help of compile time magic. Since these allocations are directly influenced by what user sends us, we should make sure we are not allocating too much to be able to fit the incoming user data.